### PR TITLE
Refactoring GitHub metrics

### PR DESCRIPTION
**Pull request by lucyb**
I've refactored the metrics code to make some future changes easier. Most of the changes just involve moving code between modules. However, there are a couple of functional changes along the way, these are:

1. Ensuring all PR data is downloaded before the old table is dropped. This provides a nicer experience if there's an issue with the GitHub API
2. Avoiding fetching PR data for repos that are on the deny list. This should provide a performance increase as it reduces the number of GraphQL queries that we're doing.

I've structured the commits so that you can see the refactoring steps I took and tried to explain my reasoning for the less obvious ones.

### CODE REVIEW

File: tests/metrics/github/test_security.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def fake_vulnerabilities(client, repo):
 
 
 def test_get_repos():
-    security.query_repos = fake_repos
+    security.query.repos = fake_repos
```
Comment (iaindillingham) [inline tests/metrics/github/test_security.py:65]: Do you know why we're not using pytest's `monkeypatch` fixture, here? It took me awhile to understand why we're accessing `repos` (or, previously, `query_repos`) through `security` and not through `query`.

**Review by benbc — APPROVED** [review]
I think my earlier comment is pretty much obsoleted by your later changes, so don't feel you need to do anything about it.

This is all great. Thank you.

File: metrics/github/cli.py [hunk 2]
```diff
@@ -169,6 +169,11 @@ def tech_owned_repo(pr):
     return not (pr["org"] in NON_TECH_REPOS and pr["repo"] in NON_TECH_REPOS[pr["org"]])
 
 
+def fetch_prs(client):
```
Comment (iaindillingham) [inline metrics/github/cli.py]: I was confused by when we use `iter_*` and when we use `fetch_*`, but having read through the PR I don't think there's a convention, as such (both signal generator functions). Have I misunderstood?

**Review by iaindillingham — APPROVED** [review]
Thank you for breaking this PR into stages. As ever, doing so makes it much easier to step through the commits, check out locally if necessary, and figure out what's going on. I've made several comments, but otherwise this looks good to me.

File: metrics/github/api.py [hunk 3]
```diff
@@ -137,7 +137,6 @@ def iter_repo_prs(client, repo):
     for pr in client.get_query(
         query,
         path=["organization", "repository", "pullRequests"],
-        org=client.org,
         repo=repo["name"],
```
[the remark this answers is not in the document]
Comment (lucyb) [reply metrics/github/api.py]: In this query, we're fetching the PRs for a given repo, so it is required. The GraphQL query specifies the parameters that it needs here:
```
query prs($cursor: String, $org: String!, $repo: String!) {
```

The org can be inferred from the client object, but the repo needs to be passed in.

File: metrics/github/cli.py (hunk 2 above)
Comment (lucyb) [reply metrics/github/cli.py]: That's a good spot. We've already used `iter_prs` in prs.py, which is where this function ended up, so I needed a different name. I think there's probably a better name for that function though or some further refactoring needed to stop us from wanting to call two separate functions the same thing.

File: tests/metrics/github/test_security.py (hunk 1 above)
Comment (lucyb) [reply tests/metrics/github/test_security.py]: That's a good question. I tend to prefer doing it this way, but I've also had feedback from others that it's easier to understand what's happening by using `monkeypatch`. Anyway, I've taken the hint and changed it. It's more consistent with the other tests now too.

File: metrics/github/prs.py [hunk 4]
```diff
@@ -210,12 +210,13 @@ def pr_throughput(prs):
 }
 
 
-def tech_owned_repo(pr):
+def tech_owned_repo(repo, org):
```
[the remark this answers is not in the document]
Comment (lucyb) [reply metrics/github/prs.py]: Good suggestion, thanks. That's been pushed.

**Pull request merged**